### PR TITLE
Rancher Monitoring Support

### PR DESCRIPTION
**Pull request by orangedeng**
Related Issue: https://github.com/rancher/rancher/issues/14230
Related PR: https://github.com/rancher/types/pull/629

This is a PR for rancher monitoring support. Summaries are follow:
1. Support prometheus crds in Rancher API.
2. Adding new kind of Alert feature that supports grouping alert rules.
3. Enable and disable cluster and project monitoring.
4. Support monitoring metrics graph.
5. Support prometheus to collect workload metrics with operator

### CODE REVIEW

File: apis/management.cattle.io/v3/authz_types.go [hunk 1]
```diff
@@ -52,6 +54,7 @@ type ProjectSpec struct {
 	ClusterName                   string                  `json:"clusterName,omitempty" norman:"required,type=reference[cluster]"`
 	ResourceQuota                 *ProjectResourceQuota   `json:"resourceQuota,omitempty"`
 	NamespaceDefaultResourceQuota *NamespaceResourceQuota `json:"namespaceDefaultResourceQuota,omitempty"`
+	EnableProjectMonitoring       *bool                   `json:"enableProjectMonitoring" norman:"default=false"`
```
Comment (ibuildthecloud) [inline apis/management.cattle.io/v3/authz_types.go]: why is this a pointer?

File: apis/management.cattle.io/v3/cluster_types.go [hunk 2]
```diff
@@ -74,6 +76,7 @@ type ClusterSpec struct {
 	DefaultClusterRoleForProjectMembers  string                               `json:"defaultClusterRoleForProjectMembers,omitempty" norman:"type=reference[roleTemplate]"`
 	DockerRootDir                        string                               `json:"dockerRootDir,omitempty" norman:"default=/var/lib/docker"`
 	EnableNetworkPolicy                  *bool                                `json:"enableNetworkPolicy" norman:"default=false"`
+	EnableClusterMonitoring              *bool                                `json:"enableClusterMonitoring" norman:"default=false"`
```
Comment (ibuildthecloud) [inline apis/management.cattle.io/v3/cluster_types.go]: why is this a pointer also

File: apis/management.cattle.io/v3/alerting_types.go [hunk 3]
```diff
@@ -99,6 +99,155 @@ type AlertStatus struct {
 	AlertState string `json:"alertState,omitempty" norman:"options=active|inactive|alerting|muted,default=active"`
 }
 
+type ClusterAlertGroup struct {
```
Comment (cjellick) [inline apis/management.cattle.io/v3/alerting_types.go]: Where are all these types used? Are there rancher controllers that act on these?

File: apis/management.cattle.io/v3/authz_types.go (hunk 1 above)
Comment (thxCode) [reply apis/management.cattle.io/v3/authz_types.go]: similar to `Cluster` struct.

File: apis/management.cattle.io/v3/cluster_types.go (hunk 2 above)
Comment (thxCode) [reply apis/management.cattle.io/v3/cluster_types.go]: similar to `EnableNetworkPolicy`, keep them in the same style.

File: apis/management.cattle.io/v3/authz_types.go (hunk 1 above)
Comment (cjellick) [reply apis/management.cattle.io/v3/authz_types.go]: We only do pointer bools when we _need_ to. That generally is to support backwards compatibility and upgrade scenarios. If it doesn't have a functional need to be a pointer, it shouldn't be.

**Comment by cjellick** [comment]
Regarding the alert group changes, are you introducing any breaking API changes like removing or renaming a field? As best I can tell from reviewing, you are not, but I wanted to ask the question explicitly for your consideration.

File: apis/management.cattle.io/v3/schema/schema.go [hunk 4]
```diff
@@ -501,12 +512,8 @@ func alertTypes(schema *types.Schemas) *types.Schemas {
 	return schema.
 		AddMapperForType(&Version, v3.Notifier{},
 			m.DisplayName{}).
-		AddMapperForType(&Version, v3.ClusterAlert{},
-			&m.Embed{Field: "status"},
-			m.DisplayName{}).
-		AddMapperForType(&Version, v3.ProjectAlert{},
-			&m.Embed{Field: "status"},
-			m.DisplayName{}).
+		MustImport(&Version, v3.ClusterAlert{}).
```
Comment (cjellick) [inline apis/management.cattle.io/v3/schema/schema.go]: why did you drop the embedding of status and the displayName mapper here?

File: apis/management.cattle.io/v3/cluster_types.go [hunk 5]
```diff
@@ -302,3 +306,7 @@ type IngressCapabilities struct {
 	IngressProvider      string `json:"ingressProvider,omitempty"`
 	CustomDefaultBackend bool   `json:"customDefaultBackend,omitempty"`
 }
+
+type MonitoringInput struct {
+	Answers map[string]string `json:"answers,omitempty"`
```
Comment (cjellick) [inline apis/management.cattle.io/v3/cluster_types.go]: What is the UX going to be around this? How are we going to present the questions and answers?

File: apis/management.cattle.io/v3/authz_types.go (hunk 1 above)
Comment (thxCode) [reply apis/management.cattle.io/v3/authz_types.go]: @cjellick , Got.

File: apis/management.cattle.io/v3/cluster_types.go (hunk 5 above)
Comment (thxCode) [reply apis/management.cattle.io/v3/cluster_types.go]: This struct uses to overwrite the Answers of `cluster-monitoring` App and `project-monitoring` App, and we only allow users to overwrite some important fields, just like Prometheus retention, Storage Class, Storage Size and Node Exporter Port.

**Comment by cjellick** [comment]
LGTM

**Pull request merged**